### PR TITLE
Better Number Calculation

### PR DESCRIPTION
**Pull request by RafaelLeeImg**
Details are wrote in issue 948
https://github.com/stlink-org/stlink/issues/948

### CODE REVIEW

**Comment by Nightwalker-87** [comment]
@RafaelLeeImg: Please make sure that all travis CI test pass successfully.

**Comment by Nightwalker-87** [comment]
I'm afraid to say this still does not seem to qualify for a merge yet...
Please ensure there are no errors left.

**Comment by chenguokai** [comment]
During my local test `make test` failure with the following error:

```
"test-flash" start time: Apr 30 12:24 CST
Output:
----------------------------------------------------------
*** Error: Integer greater than INT32_MAX, cannot convert to int32_t
*** Error: Invalid value for addr
*** Error: Integer greater than INT32_MAX, cannot convert to int32_t
*** Error: Invalid value for addr
*** Error: Integer greater than INT32_MAX, cannot convert to int32_t
*** Error: Invalid value for addr
*** Error: Integer greater than INT32_MAX, cannot convert to int32_t
*** Error: Invalid value for addr
*** Error: Expected args for this command: write <path> <addr>
*** Error: Expected args for this command: write <path>
*** Error: Invalid value for addr
[OK] (-1) 
[ERROR] (-1) --debug --reset read test.bin 0x80000000 0x1000
[ERROR] (-1) --debug --reset write test.bin 0x80000000
[ERROR] (-1) --debug --reset read test.bin 0x80000000 0x1000
[ERROR] (-1) --debug --reset write test.bin 0x80000000
[OK] (0) erase
[OK] (0) --debug --reset --format=ihex write test.hex
[OK] (-1) --debug --reset --format=binary write test.hex
[OK] (-1) --debug --reset --format=ihex write test.hex 0x80000000
[OK] (-1) --debug --reset write test.hex sometext
[OK] (-1) --serial A1020304 erase sometext
[OK] (0) --serial A1020304 erase
[OK] (0) --serial=A1020304 erase
<end of output>
Test time =   0.00 sec

```

I think the four failed test should be passed before merge.
It would be better if you are kind enough to add more tests for your new input format support.

**Comment by Nightwalker-87** [comment]
@RafaelLeeImg: It does not work this way. Please resolve this and work it out on your local branch first before pushing any further changes. If not capable of doing so, please ask for assistance from the community in the linked issue. An open PR should not be a place for trial and error tests. If the next commit of this PR does not resolve the build failures, I may decline it without further notice.  Thank you for your understanding.

**Comment by RafaelLeeImg** [comment]
Sorry, I didn't notice there are test cases in the project before, there's no description about local test in README.md or contribution guidelines.

Now CI fixed, local test cases added and local test passed.
I've read the contribution guidelines.
I've tried to compile the code with clang 7.0.1-8 as well as gcc 8.3.0-6, on my local computer, no error reported, I've checked the CFLAGS of -Wall -Wextra are all set. And I tried the with real devices, no error reported. The I made the pull request only found CI failed.

File: tests/flash.c [hunk 1]
```diff
@@ -95,16 +95,36 @@ static struct Test tests[] = {
           .log_level = DEBUG_LOG_LEVEL,
           .format = FLASH_FORMAT_BINARY }
     },
-    { "--debug --reset read test.bin 0x80000000 0x1000", 0,
```
[the remark this answers is not in the document]
Comment (RafaelLeeImg) [reply tests/flash.c:98]: The original block is duplicated with the one 2 blocks above. So I replaced it with the one I wanted to add. If you read the whole file you will see that.

**Comment by Nightwalker-87** [comment]
@RafaelLeeImg: This looks a lot better now. Thank you. :+1: 
Indeed the test files are not mentioned as I could verify by myself now - they should however, as they have been present for quite a long time now.
I'll denote them in one of the upcoming commits.

**Pull request merged**